File: leetcode/cpp/0007-reverse-integer.cpp
/** https://leetcode.com/problems/reverse-integer/
 *
 * Reverse Integer
 *
 *
 */

#include <climits>

class Solution {
public:
    int reverse(int x) {
        int rev{0};

        while (x != 0) {
            if (rev > INT_MAX / 10 || rev < INT_MIN / 10) {
                return 0;
            }

            rev = rev * 10 + x % 10;

            x /= 10;
        }

        return rev;
    }
};

int main() {
    Solution().reverse(123);          // 321
    Solution().reverse(-123);         // -321
    Solution().reverse(120);          // 21
    Solution().reverse(1534236469);   // 0
    Solution().reverse(-2147483648);  // 0

    return 0;
}

File: leetcode/typescript/0007-reverse-integer.ts
/** https://leetcode.com/problems/reverse-integer/
 *
 * Reverse Integer
 *
 * @param {number} x
 * @returns {number}
 */
function reverse(x: number): number {
  let y = Math.abs(x);
  let rev = 0;
  const MAX_32_BIT_SIGNED_INT = 2 ** 31 - 1;

  while (y > 0) {
    rev = rev * 10 + (y % 10);

    if (rev > MAX_32_BIT_SIGNED_INT) {
      return 0;
    }

    y = Math.floor(y / 10);
  }

  return x > 0 ? rev : rev * -1;
}

console.log(reverse(123)); // 321
console.log(reverse(-123)); // -321
console.log(reverse(120)); // 21
console.log(reverse(1534236469)); // 0
